packages/@react-stately/data/src/useAsyncList.ts

```diff
@@ -243,34 +272,50 @@ export function useAsyncList<T, C = string>(options: AsyncListOptions<T, C>): As
         selectedKeys: data.selectedKeys,
         sortDescriptor: action.sortDescriptor ?? data.sortDescriptor,
         signal: abortController.signal,
-        cursor: action.type === 'loadingMore' ? data.cursor : null
+        cursor: action.type === 'loadingMore' ? data.cursor : null,
+        filterText: data.filterText
       });
       dispatch({type: 'success', ...response, abortController});
     } catch (e) {
       dispatch({type: 'error', error: e, abortController});
     }
   };
 
+  // Handle initial load
   useEffect(() => {
     dispatchFetch({type: 'loading'}, load);
   // eslint-disable-next-line react-hooks/exhaustive-deps
   }, []);
 
+  // Reload on changes to filter text but only if performing server side filtering
+  // Kinda weird since it will trigger a reload when user selects a option
+  useEffect(() => {
+    if (!filterFn) {
+      dispatchFetch({type: 'loading'}, load);
+    }
+  }, [data.filterText, filterFn, load, dispatchFetch]);
+
+  let filteredItems = useMemo(
+    () => filterFn ? data.items.filter(item => filterFn(item, data.filterText)) : data.items,
+    [data.items, data.filterText, filterFn]);
+
   return {
-    items: data.items,
+    items: filteredItems,
     selectedKeys: data.selectedKeys,
     sortDescriptor: data.sortDescriptor,
     isLoading: data.state === 'loading' || data.state === 'loadingMore' || data.state === 'sorting',
     error: data.error,
+    filterText: data.filterText,
     getItem(key: Key) {
       return data.items.find(item => getKey(item) === key);
     },
     reload() {
       dispatchFetch({type: 'loading'}, load);
     },
     loadMore() {
-      // Ignore if already loading more.
-      if (data.state === 'loadingMore' || data.cursor == null) {
+      // Ignore if already loading or loading more.
+      // TODO: ask if this is appropriate
+      if (data.state === 'loadingMore' || data.state === 'loading' || data.cursor == null) {
```

Comment:
If the user triggers a brand new filter load by typing when the dropdown is open, loadMore actually gets called by listbox (due to scroll position? not super sure will need to dig) and thus clears the item list due to the `loadingMore` -> `loading` case above. I figure we should early return anyways if `loadingMore` is triggered when `loading` is happening anyway

Reply:
I think if you use a new action type for filtering that behaves closer to sorting/loading more and doesn't clear it might solve this